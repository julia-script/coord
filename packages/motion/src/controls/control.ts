import { tween } from "@/tweening";
import { Thread } from "@/utils";
import {
  EasingOptions,
  isFunction,
} from "@coord/core";

export type Dispatcher<
  TValue,
  TValueIn = TValue
> = TValueIn | ((prev: TValue) => TValueIn);
export class Control<TValue, TValueIn = TValue> {
  private _deferred: Dispatcher<
    TValue,
    TValueIn
  >[] = [];

  protected _computeDeferred(
    deferred: Dispatcher<TValue, TValueIn>[]
  ) {
    for (const value of deferred) {
      this.set(value);
    }

    return this.get();
  }
  protected _applyDeferred<
    TThread extends Thread
  >(
    fn: (next: () => TValue) => TThread
  ): TThread {
    const deferred = this._deferred;
    this._deferred = [];
    return fn(() =>
      this._computeDeferred(deferred)
    );
  }

  protected _defer(
    value: Dispatcher<TValue, TValueIn>
  ) {
    this._deferred.push(value);
  }

  constructor(
    public _get: () => TValue,
    public _set: (value: TValueIn) => void
  ) {}
  get = () => this._get();

  set = (value: Dispatcher<TValue, TValueIn>) => {
    this._set(
      isFunction(value)
        ? value(this.get())
        : value
    );
  };
  *as(value: Dispatcher<TValue, TValueIn>) {
    this.set(value);
    yield;
  }

  tween(
    duration: number,
    fn: (
      t: number,
      initialValue: TValue
    ) => TValueIn,
    easing?: EasingOptions
  ) {
    const self = this;
    function* gen(next: () => TValue) {
      const from = next();
      yield* tween(
        duration,
        (t) => self.set(fn(t, from)),
        easing
      );
    }
    return this._applyDeferred(gen);
  }
}

export function control<TValue>(
  getter: () => TValue,
  setter: (value: TValue) => void
) {
  return new Control<TValue>(getter, setter);
}
