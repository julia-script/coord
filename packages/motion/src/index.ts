export * from "./player";
export * from "./utils";
export * from "./flow";
export * from "./tweening";
export * from "./controls";
export * from "./effects";
export * from "./motion";
