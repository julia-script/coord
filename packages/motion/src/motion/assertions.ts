import {
  YieldedType,
  isObject,
} from "@coord/core";
import {
  ContextRequest,
  EndRequest,
  MakeStateRequest,
  MotionRequest,
  MotionState,
} from "./types";

export const isRequest = (
  value: unknown
): value is MotionRequest => {
  return isObject(value) && "type" in value;
};

export const isContextRequest = <
  TState extends MotionState
>(
  value: unknown
): value is YieldedType<
  ContextRequest<TState>
> => {
  return (
    isRequest(value) &&
    value.type === "REQUEST_CONTEXT"
  );
};

export const isMakeStateRequest = <
  TKey extends string,
  TValue
>(
  value: unknown
): value is MakeStateRequest<TKey, TValue> => {
  return (
    isRequest(value) &&
    value.type === "MAKE_STATE" &&
    "state" in value &&
    isObject(value.state) &&
    "key" in value.state &&
    "initialState" in value.state
  );
};

export const isEndRequest = (
  value: unknown
): value is EndRequest => {
  return (
    isRequest(value) &&
    value.type === "REQUEST_END"
  );
};
