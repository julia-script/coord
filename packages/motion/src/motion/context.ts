import {
  isContextRequest,
  isEndRequest,
  isMakeStateRequest,
  isRequest,
} from "./assertions";
import {
  MotionBuilder,
  MotionMeta,
  MotionRequest,
  MotionScene,
  MotionSettings,
  MotionState,
  SceneState,
} from "./types";

type FullState<TState extends MotionState> = TState & {
  $transition: number;
  $frame: number;
  $done: boolean;
};
type FullStateKeys<TState extends MotionState> = keyof FullState<TState> &
  string;

export class MotionContext<TState extends MotionState> {
  _state: FullState<TState>;
  frames: FullState<TState>[] = [];
  meta: MotionMeta = {
    title: "Untitled",
    start: 0,
    duration: 0,
    scenes: {},
  };

  settings: MotionSettings = {
    fps: 60,
    physicsFps: 120,
  };

  get time() {
    return this.frames.length / this.settings.fps;
  }
  constructor() {
    this._state = {} as FullState<TState>;
  }

  static from<TScene extends MotionScene<MotionBuilder>>(
    scene: TScene,
    settings: Partial<MotionSettings> & { transitionDuration?: number } = {}
  ): MotionContext<SceneState<TScene>> {
    const context = new MotionContext<SceneState<TScene>>();
    context.meta.title = scene.meta.title;
    context.meta.description = scene.meta.description;

    const { transitionDuration = 0, ...rest } = settings;
    context.settings = { ...context.settings, ...rest };
    const builderIterator = scene.builder();

    let transitionDurationInFrames = context.settings.fps * transitionDuration;
    context._state.$done = false;

    let done = false;
    while (true) {
      if (done === false) {
        const currentIteration = builderIterator.next();
        done = !!currentIteration.done;
        if (isRequest(currentIteration.value)) {
          context.respondRequest(currentIteration.value);
          continue;
        }
      }
      if (
        done === true &&
        context.frames.length >= transitionDurationInFrames
      ) {
        break;
      }

      context._state.$frame = context.frames.length;

      context._state.$transition =
        transitionDurationInFrames <= 0
          ? 1
          : Math.min(
              (context.frames.length + 1) / transitionDurationInFrames,
              1
            );

      context.pushFrame();
    }
    if (context.frames.length === 0) {
      context._state.$done = true;
      context.pushFrame();
    }
    return context;
  }

  pushFrame() {
    this.meta.duration++;
    this.frames.push({
      ...this._state,
    });
  }

  setState<TKey extends FullStateKeys<TState>>(
    key: TKey,
    value: (typeof this._state)[TKey]
  ) {
    this._state[key] = value;
  }

  respondRequest<TRequest extends MotionRequest>(request: TRequest) {
    if (isContextRequest<TState>(request)) {
      request.context = this;
    } else if (
      isMakeStateRequest<
        FullStateKeys<TState>,
        FullState<TState>[FullStateKeys<TState>]
      >(request)
    ) {
      this.makeState(request.state.key, request.state.initialState);
    } else if (isEndRequest(request)) {
      this._state.$done = true;
    }
  }

  makeState<TKey extends FullStateKeys<TState>>(
    key: TKey,
    value: FullState<TState>[TKey]
  ) {
    this.setState(key, value);

    for (let i = this.frames.length - 1; i >= 0; i--) {
      const frame = this.frames[i]!;
      if (key in frame) {
        break;
      }
      Object.assign(frame, { [key]: value });
    }
  }
}
