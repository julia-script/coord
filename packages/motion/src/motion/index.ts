export * from "./assertions";
export * from "./context";
export * from "./requests";
export * from "./types";
export * from "./make-scene";
export * from "./run-motion";
