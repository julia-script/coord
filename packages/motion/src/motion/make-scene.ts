import { SceneMetaish, generateMeta } from "@/utils";
import { MotionBuilder } from "./types";

export function makeMotion<TBuilder extends MotionBuilder>(
  meta: SceneMetaish,
  builder: TBuilder
) {
  return {
    meta: generateMeta(meta),
    builder,
  };
}
