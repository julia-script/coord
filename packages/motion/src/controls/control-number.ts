import {
  EasingOptions,
  isFunction,
} from "@coord/core";
import { Control, Dispatcher } from "./control";
import {
  SpringParameters,
  spring,
  tween,
} from "@/tweening";
import { makeState } from "@/motion";

export class NumberControl extends Control<number> {
  to(value: Dispatcher<number>) {
    this._defer(value);
    return this;
  }
  in(duration: number, easing?: EasingOptions) {
    const self = this;
    return this._applyDeferred(function* (next) {
      const from = self.get();
      const to = next();

      yield* tween(
        duration,
        (t) => self.set(from + (to - from) * t),
        easing
      );
    });
  }
  spring(parameters?: SpringParameters) {
    const self = this;
    return this._applyDeferred(function* (next) {
      const from = self.get();
      const to = next();
      yield* spring(
        from,
        to,
        self.set,
        parameters
      );
    });
  }
}

export function controlNumber(
  getter: () => number,
  setter: (value: number) => void
) {
  return new NumberControl(getter, setter);
}

export function* makeNumberState<
  TKey extends string
>(key: TKey, initialState: number) {
  const control = yield* makeState(
    key,
    initialState
  );

  return new NumberControl(
    control._get,
    control._set
  );
}
