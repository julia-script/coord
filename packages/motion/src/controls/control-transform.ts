import { EasingOptions, Transform, isFunction } from "@coord/core";
import { Control, Dispatcher } from "./control";
import { SpringParameters, spring, tween } from "@/tweening";
import { makeState } from "@/motion";

export class TransformControl extends Control<Transform> {
  to(value: Dispatcher<Transform>) {
    this._defer(value);
    return this;
  }

  translateTo(x: number, y: number) {
    return this.to((t) => t.setPosition([x, y]));
  }
  scaleTo(x: number, y: number = x) {
    return this.to((t) => t.setScale([x, y]));
  }

  rotateTo(angle: number) {
    return this.to((t) => t.setRotation(angle));
  }

  translate(x: number, y: number) {
    return this.to((t) => t.translate([x, y]));
  }

  scale(x: number, y: number = x) {
    return this.to((t) => t.scale([x, y]));
  }

  rotate(angle: number) {
    return this.to((t) => t.rotate(angle));
  }

  in(duration: number, easing?: EasingOptions) {
    const self = this;
    return this._applyDeferred(function* (next) {
      const from = self.get();
      const to = next();
      yield* tween(duration, (t) => self.set(from.lerp(to, t)), easing);
    });
  }
}

export function controlTransform(
  getter: () => Transform,
  setter: (value: Transform) => void
) {
  return new Control(getter, setter);
}

export function* makeTransformState<TKey extends string>(
  key: TKey,
  initialState: Transform
) {
  const control = yield* makeState(key, initialState);

  return new TransformControl(control._get, control._set);
}
