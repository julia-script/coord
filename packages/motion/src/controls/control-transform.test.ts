import { makeState, runMotion } from "@/motion";
import { makeMotion } from "@/motion";
import { test, describe, expect } from "vitest";
import { control } from "./control";
import { makeNumberState } from "./control-number";
import { makeTransformState } from "./control-transform";
import { transform } from "@coord/core";

describe("TransformControl", async () => {
  test("tweening", async () => {
    const scene = makeMotion(
      "Test",
      function* () {
        const a = yield* makeTransformState(
          "a",
          transform()
        );
        yield* a.translate(10, 10).scale(2).in(1);
      }
    );
    const executed = runMotion(scene);

    expect(executed.frames.length).toBe(60);
    expect(executed.frames).toMatchSnapshot();
  });
});
