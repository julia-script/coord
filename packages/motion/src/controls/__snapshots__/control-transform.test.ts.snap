// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransformControl > tweening 1`] = `
[
  {
    "$done": false,
    "$frame": 0,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0000185185185184,
        0,
        0.00018518518518518518,
        0,
        1.0000185185185184,
        0.00018518518518518518,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 1,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0001481481481482,
        0,
        0.0014814814814814814,
        0,
        1.0001481481481482,
        0.0014814814814814814,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 2,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0005,
        0,
        0.005000000000000001,
        0,
        1.0005,
        0.005000000000000001,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 3,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0011851851851852,
        0,
        0.011851851851851851,
        0,
        1.0011851851851852,
        0.011851851851851851,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 4,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0023148148148149,
        0,
        0.023148148148148147,
        0,
        1.0023148148148149,
        0.023148148148148147,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 5,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.004,
        0,
        0.04000000000000001,
        0,
        1.004,
        0.04000000000000001,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 6,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.006351851851852,
        0,
        0.06351851851851853,
        0,
        1.006351851851852,
        0.06351851851851853,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 7,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0094814814814814,
        0,
        0.09481481481481481,
        0,
        1.0094814814814814,
        0.09481481481481481,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 8,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0135,
        0,
        0.135,
        0,
        1.0135,
        0.135,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 9,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0185185185185186,
        0,
        0.18518518518518517,
        0,
        1.0185185185185186,
        0.18518518518518517,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 10,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0246481481481482,
        0,
        0.2464814814814814,
        0,
        1.0246481481481482,
        0.2464814814814814,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 11,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.032,
        0,
        0.32000000000000006,
        0,
        1.032,
        0.32000000000000006,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 12,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0406851851851853,
        0,
        0.4068518518518519,
        0,
        1.0406851851851853,
        0.4068518518518519,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 13,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0508148148148149,
        0,
        0.5081481481481482,
        0,
        1.0508148148148149,
        0.5081481481481482,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 14,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0625,
        0,
        0.625,
        0,
        1.0625,
        0.625,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 15,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0758518518518518,
        0,
        0.7585185185185185,
        0,
        1.0758518518518518,
        0.7585185185185185,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 16,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.0909814814814816,
        0,
        0.9098148148148146,
        0,
        1.0909814814814816,
        0.9098148148148146,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 17,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.108,
        0,
        1.08,
        0,
        1.108,
        1.08,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 18,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.1270185185185184,
        0,
        1.270185185185185,
        0,
        1.1270185185185184,
        1.270185185185185,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 19,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.1481481481481481,
        0,
        1.4814814814814814,
        0,
        1.1481481481481481,
        1.4814814814814814,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 20,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.1715,
        0,
        1.7149999999999996,
        0,
        1.1715,
        1.7149999999999996,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 21,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.1971851851851851,
        0,
        1.9718518518518513,
        0,
        1.1971851851851851,
        1.9718518518518513,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 22,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.2253148148148147,
        0,
        2.2531481481481483,
        0,
        1.2253148148148147,
        2.2531481481481483,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 23,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.256,
        0,
        2.5600000000000005,
        0,
        1.256,
        2.5600000000000005,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 24,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.2893518518518519,
        0,
        2.893518518518519,
        0,
        1.2893518518518519,
        2.893518518518519,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 25,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.3254814814814815,
        0,
        3.2548148148148153,
        0,
        1.3254814814814815,
        3.2548148148148153,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 26,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.3645,
        0,
        3.6450000000000005,
        0,
        1.3645,
        3.6450000000000005,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 27,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.4065185185185185,
        0,
        4.065185185185186,
        0,
        1.4065185185185185,
        4.065185185185186,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 28,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.451648148148148,
        0,
        4.516481481481481,
        0,
        1.451648148148148,
        4.516481481481481,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 29,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.5,
        0,
        5,
        0,
        1.5,
        5,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 30,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.548351851851852,
        0,
        5.48351851851852,
        0,
        1.548351851851852,
        5.48351851851852,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 31,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.5934814814814815,
        0,
        5.934814814814815,
        0,
        1.5934814814814815,
        5.934814814814815,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 32,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.6355,
        0,
        6.355,
        0,
        1.6355,
        6.355,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 33,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.6745185185185185,
        0,
        6.745185185185186,
        0,
        1.6745185185185185,
        6.745185185185186,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 34,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.7106481481481484,
        0,
        7.106481481481483,
        0,
        1.7106481481481484,
        7.106481481481483,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 35,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.744,
        0,
        7.4399999999999995,
        0,
        1.744,
        7.4399999999999995,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 36,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.7746851851851853,
        0,
        7.7468518518518525,
        0,
        1.7746851851851853,
        7.7468518518518525,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 37,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.8028148148148149,
        0,
        8.028148148148148,
        0,
        1.8028148148148149,
        8.028148148148148,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 38,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.8285,
        0,
        8.285,
        0,
        1.8285,
        8.285,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 39,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.8518518518518519,
        0,
        8.518518518518519,
        0,
        1.8518518518518519,
        8.518518518518519,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 40,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.8729814814814816,
        0,
        8.729814814814814,
        0,
        1.8729814814814816,
        8.729814814814814,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 41,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.892,
        0,
        8.919999999999998,
        0,
        1.892,
        8.919999999999998,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 42,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9090185185185184,
        0,
        9.090185185185186,
        0,
        1.9090185185185184,
        9.090185185185186,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 43,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9241481481481482,
        0,
        9.241481481481483,
        0,
        1.9241481481481482,
        9.241481481481483,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 44,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9375,
        0,
        9.375,
        0,
        1.9375,
        9.375,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 45,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9491851851851854,
        0,
        9.491851851851852,
        0,
        1.9491851851851854,
        9.491851851851852,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 46,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9593148148148147,
        0,
        9.593148148148149,
        0,
        1.9593148148148147,
        9.593148148148149,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 47,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.968,
        0,
        9.68,
        0,
        1.968,
        9.68,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 48,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9753518518518518,
        0,
        9.753518518518518,
        0,
        1.9753518518518518,
        9.753518518518518,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 49,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9814814814814814,
        0,
        9.814814814814815,
        0,
        1.9814814814814814,
        9.814814814814815,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 50,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9865,
        0,
        9.865,
        0,
        1.9865,
        9.865,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 51,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9905185185185186,
        0,
        9.905185185185186,
        0,
        1.9905185185185186,
        9.905185185185186,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 52,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9936481481481483,
        0,
        9.936481481481481,
        0,
        1.9936481481481483,
        9.936481481481481,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 53,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.996,
        0,
        9.96,
        0,
        1.996,
        9.96,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 54,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9976851851851851,
        0,
        9.976851851851853,
        0,
        1.9976851851851851,
        9.976851851851853,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 55,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9988148148148148,
        0,
        9.988148148148149,
        0,
        1.9988148148148148,
        9.988148148148149,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 56,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9994999999999998,
        0,
        9.995,
        0,
        1.9994999999999998,
        9.995,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 57,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9998518518518518,
        0,
        9.99851851851852,
        0,
        1.9998518518518518,
        9.99851851851852,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 58,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        1.9999814814814814,
        0,
        9.999814814814815,
        0,
        1.9999814814814814,
        9.999814814814815,
        0,
        0,
        1,
      ],
    },
  },
  {
    "$done": false,
    "$frame": 59,
    "$transition": 1,
    "a": Transform {
      "_matrix": [
        2,
        0,
        10,
        0,
        2,
        10,
        0,
        0,
        1,
      ],
    },
  },
]
`;
